### PR TITLE
[RenderEditable] Dont paint caret when selection is invalid

### LINKED ISSUE #79495: `RenderEditable` should not show the caret when the selection is invalid (i.e. `(-1, -1)`)
This corresponds to a TODO item:

https://github.com/flutter/flutter/blob/f86c859bea40892cd50784673d8c9f3a28ee3b6d/packages/flutter/lib/src/rendering/editable.dart#L3691-L3694

### PR DESCRIPTION
**Pull request by LongCatIsLooong**
Fixes https://github.com/flutter/flutter/issues/79495

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/test/rendering/editable_test.dart [hunk 1]
```diff
@@ -412,6 +412,52 @@ void main() {
...
+    );
+
+    // Let the RenderEditable paint again. Setting the selection to null should
+    // prevent the caret from being painted.
+    editable.selection = null;
+    // Still paints the paragraph.
+    expect(editable, paints..paragraph());
+    // No longer paints the caret.
+    expect(editable, isNot(paints..rect(color: Colors.red[500])),
+    );
```
Comment (justinmc) [inline packages/flutter/test/rendering/editable_test.dart:458]: Nit: This parenthesis should be on the previous line I think.

**Review by justinmc — APPROVED** [review]
LGTM 👍 

Maybe run the Google tests on this before merge?  If you think there's any risk of a cursor no longer being painted in a visual test.

**Comment by LongCatIsLooong** [comment]
Not going to merge this just yet. I'll wait for https://github.com/flutter/flutter/pull/79656 to be rolled and do a TGP.

**Comment by xu-baolin** [comment]
The behavior after this patch is consistent with the native behavior(desktop and mobile), right?
I think this is probably a big breaking change, we should explain in the document, of course, this should be done if the native behavior is also the same!

**Comment by LongCatIsLooong** [comment]
@xu-baolin I have not tested on native platforms but not painting the caret seems to be the only approach that makes sense. I'll run all customer tests to figure out whether this is a breaking change. This PR depends on a breaking change #79656 so I'll have to wait for that commit to get rolled first.

**Comment by LongCatIsLooong** [comment]
It does break internal tests: https://test.corp.google.com/ui#id=OCL:366524237:BASE:367688556:1617998508009:3bd6c925(internal link). I'm creating a migration guide.

**Pull request merged**